File: L2192/2192.cpp
#include <vector>
#include <algorithm>
using namespace std;
vector<int>::iterator iter;

class Solution {
public:
    vector<vector<int>> adj;
    vector<vector<int>> ans;
    bool visited[1000];
    void dfs(int v, bool* visited) {
        if (visited[v]) {
            return;
        } else {
            visited[v] = true;
        }
        for (int i=0; i<adj[v].size(); i++) {
            int anc_node = adj[v][i];
            if (not visited[anc_node]) {
                dfs(anc_node, visited);
            };
            vector<int> tmp(1000);
            sort(ans[v].begin(), ans[v].end());
            sort(ans[anc_node].begin(), ans[anc_node].end());
            iter = set_union(ans[v].begin(), ans[v].end(), ans[anc_node].begin(), ans[anc_node].end(), tmp.begin());
            tmp.resize(iter-tmp.begin());
            ans[v] = tmp;
        }
    }

    vector<vector<int>> getAncestors(int n, vector<vector<int>>& edges) {
        // make adj (to->from)
        for (int i=0; i<n; i++) {
            vector<int> tmp;
            adj.push_back(tmp);
            ans.push_back(tmp);
        }
        for (int i=0; i<edges.size(); i++) {
            int from = edges[i][0];
            int to = edges[i][1];
            adj[to].push_back(from);
            ans[to].push_back(from);
        }
        // dfs
        for (int node=0; node<n; node++) {
            dfs(node, visited);
        }
        return ans;
    }
};

File: L2192/2192.py
from typing import List
from collections import deque


class Solution:
    def getAncestors(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        ans = []
        # transpose (to -> from)
        trans = [[] for _ in range(n)]
        for p, c in edges:
            trans[c].append(p)
        for node in range(n):
            visit = set()
            q = deque([node])
            while q:
                curr = q.popleft()
                for anc_node in trans[curr]:
                    if anc_node not in visit and anc_node not in q:
                        visit.add(anc_node)
                        q.append(anc_node)
            ans.append(sorted(list(visit)))
        return ans


